sherpa-onnx/csrc/online-recognizer-transducer-nemo-impl.h

```diff
@@ -0,0 +1,205 @@
+      : config_(config),
+        symbol_table_(mgr, config_.model_config.tokens),
+        model_(std::make_unique<OnlineTransducerNeMoModel>(
+            mgr, config_.model_config)) {
+    if (config_.decoding_method == "greedy_search") {
+      decoder_ = std::make_unique<OnlineTransducerGreedySearchNeMoDecoder>(
+          model_.get(), config_.blank_penalty);
+    } else {
+      SHERPA_ONNX_LOGE("Unsupported decoding method: %s",
+                      config_.decoding_method.c_str());
+      exit(-1);
+    }
+
+    PostInit();
+  }
+#endif
+
+  std::unique_ptr<OnlineStream> CreateStream() const override {
+    auto stream = std::make_unique<OnlineStream>(config_.feat_config);
+    stream->SetStates(model_->GetInitStates());
+    InitOnlineStream(stream.get());
+    return stream;
+  }
+
+  void DecodeStreams(OnlineStream **ss, int32_t n) const override {
+    int32_t chunk_size = model_->ChunkSize();
+    int32_t chunk_shift = model_->ChunkShift();
+
+    int32_t feature_dim = ss[0]->FeatureDim();
+
+    std::vector<OnlineTransducerDecoderResult> results(n);
+    std::vector<float> features_vec(n * chunk_size * feature_dim);
+    std::vector<std::vector<Ort::Value>> states_vec(n);
+    std::vector<int64_t> all_processed_frames(n);
+
+    for (int32_t i = 0; i != n; ++i) {
+      const auto num_processed_frames = ss[i]->GetNumProcessedFrames();
+      std::vector<float> features =
+          ss[i]->GetFrames(num_processed_frames, chunk_size);
+
+      // Question: should num_processed_frames include chunk_shift?
+      ss[i]->GetNumProcessedFrames() += chunk_shift;
+
+      std::copy(features.begin(), features.end(),
+                features_vec.data() + i * chunk_size * feature_dim);
+
+      results[i] = std::move(ss[i]->GetResult());
+      states_vec[i] = std::move(ss[i]->GetStates());
+      all_processed_frames[i] = num_processed_frames;
+    }
+
+    auto memory_info =
+        Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
+
+    std::array<int64_t, 3> x_shape{n, chunk_size, feature_dim};
+
+    Ort::Value x = Ort::Value::CreateTensor(memory_info, features_vec.data(),
+                                            features_vec.size(), x_shape.data(),
+                                            x_shape.size());
+
+    auto states = model_->StackStates(states_vec);
+    int32_t num_states = states.size();
+    auto t = model_->RunEncoder(std::move(x), std::move(states));
+    // t[0] encoder_out, float tensor, (batch_size, dim, T)
+    // t[1] encoder_out_length, int64 tensor, (batch_size,)
+    
+    std::vector<Ort::Value> out_states;
+    out_states.reserve(num_states);
+    
+    for (int32_t k = 1; k != num_states + 1; ++k) {
+      out_states.push_back(std::move(t[k]));
+    }
+
+    Ort::Value encoder_out = Transpose12(model_->Allocator(), &t[0]);
+    
+    // defined in online-transducer-greedy-search-nemo-decoder.h
+    decoder_-> Decode(std::move(encoder_out), std::move(t[1]),
```

Comment:
By the way, you don't need to pass the encoder model states to the greedy search decoder.

Please pass the decoder model states to it instead.

Please read carefully our python streaming transducer greedy search decoding example.

https://github.com/k2-fsa/sherpa-onnx/blob/master/scripts/nemo/fast-conformer-hybrid-transducer-ctc/test-onnx-transducer.py#L190

I have posted the example here one more time in case you have not read it.

Reply:
```cpp
decoder_-> Decode(std::move(encoder_out), std::move(out_states), &results, ss, n);
```
is this correct?

Reply:
No. `out_states` is from the encoder.

Remember that `out_states` is used only for the internal states of the encoder model. We don't need to use it in the greedy search decoding.

We need to pass the LSTM states from the decoder model to the greedy search decoder.

I suggest you again that you re-read the python decoding example and figure out how the decoding works.

(you need to know how LSTM works)

Reply:
working on it.

Reply:
I understand that when initializing the model in beginning, init_cache_state intializes the initial states of the encoder model. Further when decoding begins, before the first chunk is decoded, decoder model comes into action and initializes the decoder state. After a chunk has been decoded, it emits the next states for the decoder model which becomes the current states for the next chunk.

Reply:
yes, so you don't need the encoder states during greedy search decoding.

Reply:
```cpp
    for (int32_t i = 0; i != n; ++i) {
      decoder_->Decode(std::move(encoder_out), 
                      std::move(ss[i]->GetNeMoDecoderStates()), &results, ss, n);
    }
```
`GetNeMoDecoderStates` fetches the intial states of the decoder. But I am not really sure of the implementation here done above.

Reply:
please read
- (1) our python code for nemo streaming transducer greedy search decoding
- (2) our c++ code for nemo non-streaming transducer greedy search decoding

make sure you indeed understand the code.

Reply:
I revised the code
```cpp
    // defined in online-transducer-greedy-search-nemo-decoder.h
    std::vector<Ort::Value> decoder_states = model_->GetDecoderInitStates(1);
    // updated decoder states are returned
    decoder_states = decoder_->Decode(std::move(encoder_out), 
                                      std::move(decoder_states), 
                                      &results, ss, n);

    std::vector<std::vector<Ort::Value>> next_states =
        model_->UnStackStates(decoder_states);
```